x/bsonx/bsoncore/bsonbuilder.go

```diff
@@ -0,0 +1,362 @@
+// b to db.doc
+func (db *DocumentBuilder) AppendBinary(key string, subtype byte, b []byte) *DocumentBuilder {
+	db.doc = AppendBinaryElement(db.doc, key, subtype, b)
+	return db
+}
+
+// AppendBinary will append a BSON binary element using subtype, and
+// b to a.arr
+func (a *ArrayBuilder) AppendBinary(subtype byte, b []byte) *ArrayBuilder {
+	last := len(a.key) - 1
+	a.arr = AppendBinaryElement(a.arr, strconv.Itoa(a.key[last]), subtype, b)
+	a.key[last]++
+	return a
+}
+
+// AppendBoolean will append a boolean element using key and b to db.doc
+func (db *DocumentBuilder) AppendBoolean(key string, b bool) *DocumentBuilder {
+	db.doc = AppendBooleanElement(db.doc, key, b)
+	return db
+}
+
+// AppendBoolean will append a boolean element using b to a.arr
+func (a *ArrayBuilder) AppendBoolean(b bool) *ArrayBuilder {
+	last := len(a.key) - 1
+	a.arr = AppendBooleanElement(a.arr, strconv.Itoa(a.key[last]), b)
+	a.key[last]++
+	return a
+}
+
+// AppendDateTime will append a datetime element using key and dt to db.doc
+func (db *DocumentBuilder) AppendDateTime(key string, dt int64) *DocumentBuilder {
+	db.doc = AppendDateTimeElement(db.doc, key, dt)
+	return db
+}
+
+// AppendDateTime will append datetime element dt to a.arr
+func (a *ArrayBuilder) AppendDateTime(dt int64) *ArrayBuilder {
+	last := len(a.key) - 1
+	a.arr = AppendDateTimeElement(a.arr, strconv.Itoa(a.key[last]), dt)
+	a.key[last]++
+	return a
+}
+
+// AppendNull will append a null element using key to db.doc
+func (db *DocumentBuilder) AppendNull(key string) *DocumentBuilder {
+	db.doc = AppendNullElement(db.doc, key)
+	return db
+}
+
+// AppendNull will append a null element to a.arr
+func (a *ArrayBuilder) AppendNull() *ArrayBuilder {
+	last := len(a.key) - 1
+	a.arr = AppendNullElement(a.arr, strconv.Itoa(a.key[last]))
+	a.key[last]++
+	return a
+}
+
+// AppendRegex will append pattern and options using key to db.doc
+func (db *DocumentBuilder) AppendRegex(key, pattern, options string) *DocumentBuilder {
+	db.doc = AppendRegexElement(db.doc, key, pattern, options)
+	return db
+}
+
+// AppendRegex will append pattern and options to a.arr
+func (a *ArrayBuilder) AppendRegex(pattern, options string) *ArrayBuilder {
+	last := len(a.key) - 1
+	a.arr = AppendRegexElement(a.arr, strconv.Itoa(a.key[last]), pattern, options)
+	a.key[last]++
+	return a
+}
+
+// AppendJavaScript will append js using the provided key to db.doc
+func (db *DocumentBuilder) AppendJavaScript(key, js string) *DocumentBuilder {
+	db.doc = AppendJavaScriptElement(db.doc, key, js)
+	return db
+}
+
+// AppendJavaScript will append js to a.arr
+func (a *ArrayBuilder) AppendJavaScript(js string) *ArrayBuilder {
+	last := len(a.key) - 1
+	a.arr = AppendJavaScriptElement(a.arr, strconv.Itoa(a.key[last]), js)
+	a.key[last]++
+	return a
+}
+
+// AppendCodeWithScope will append code and scope using key to db.doc
+func (db *DocumentBuilder) AppendCodeWithScope(key, code string, scope []byte) *DocumentBuilder {
```

Comment:
`scope` is a document, so let's take `bsoncore.Document` here.

Reply:
I'm not sure what you mean. `AppendCodeWithScopeElement` currently accepts scope as a `[]byte` and it's probably fine to change that to take `Document`. I didn't understand your point about taking `Document` instead of string, though, because documents and strings are not interchangeable. Can you explain?

Reply:
I misunderstood. Changed scope to bsoncore.Document